In [1]:
import pandas as pd 
import numpy as np

In [3]:
# Loading the dataset 
weather_df = pd.read_csv("1. Weather Data.csv")
weather_df.head()

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   
1  1/1/2012 1:00    -1.8              -3.7         87                4   
2  1/1/2012 2:00    -1.8              -3.4         89                7   
3  1/1/2012 3:00    -1.5              -3.2         88                6   
4  1/1/2012 4:00    -1.5              -3.3         88                7   

   Visibility_km  Press_kPa               Weather  
0            8.0     101.24                   Fog  
1            8.0     101.24                   Fog  
2            4.0     101.26  Freezing Drizzle,Fog  
3            4.0     101.27  Freezing Drizzle,Fog  
4            4.8     101.23                   Fog

### Question 1. Find all records where the weather was exactly clear.

In [7]:
weather_df.Weather.unique()

array(['Fog', 'Freezing Drizzle,Fog', 'Mostly Cloudy', 'Cloudy', 'Rain',
       'Rain Showers', 'Mainly Clear', 'Snow Showers', 'Snow', 'Clear',
       'Freezing Rain,Fog', 'Freezing Rain', 'Freezing Drizzle',
       'Rain,Snow', 'Moderate Snow', 'Freezing Drizzle,Snow',
       'Freezing Rain,Snow Grains', 'Snow,Blowing Snow', 'Freezing Fog',
       'Haze', 'Rain,Fog', 'Drizzle,Fog', 'Drizzle',
       'Freezing Drizzle,Haze', 'Freezing Rain,Haze', 'Snow,Haze',
       'Snow,Fog', 'Snow,Ice Pellets', 'Rain,Haze', 'Thunderstorms,Rain',
       'Thunderstorms,Rain Showers', 'Thunderstorms,Heavy Rain Showers',
       'Thunderstorms,Rain Showers,Fog', 'Thunderstorms',
       'Thunderstorms,Rain,Fog',
       'Thunderstorms,Moderate Rain Showers,Fog', 'Rain Showers,Fog',
       'Rain Showers,Snow Showers', 'Snow Pellets', 'Rain,Snow,Fog',
       'Moderate Rain,Fog', 'Freezing Rain,Ice Pellets,Fog',
       'Drizzle,Ice Pellets,Fog', 'Drizzle,Snow', 'Rain,Ice Pellets',
       'Drizzle,Snow,Fog', 

In [6]:
# Filtering for exactly "Clear" weather records 
clear_weather_df = weather_df[weather_df.Weather == "Clear"] 
clear_weather_df.head()

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
67   1/3/2012 19:00   -16.9             -24.8         50               24   
114  1/5/2012 18:00    -7.1             -14.4         56               11   
115  1/5/2012 19:00    -9.2             -15.4         61                7   
116  1/5/2012 20:00    -9.8             -15.7         62                9   
117  1/5/2012 21:00    -9.0             -14.8         63               13   

     Visibility_km  Press_kPa Weather  
67            25.0     101.74   Clear  
114           25.0     100.71   Clear  
115           25.0     100.80   Clear  
116           25.0     100.83   Clear  
117           25.0     100.83   Clear

### Question 2. Find the number of times the wind speed was exactly 4 km/hr.

In [15]:
# Filter out data where wind speed is 4
speed_4km_per_hr_df = weather_df[weather_df["Wind Speed_km/h"] == 4]

num_of_times = len(speed_4km_per_hr_df)
num_of_times

474

- There was 474 times when the weed of the speed was exactly 4 KM/Hr

### Question 3. Check if there are any NULL values present in the dataset.

In [16]:
weather_df.isnull().sum()

Date/Time           0
Temp_C              0
Dew Point Temp_C    0
Rel Hum_%           0
Wind Speed_km/h     0
Visibility_km       0
Press_kPa           0
Weather             0
dtype: int64

- There are no null values in the dataset

### Question 4: Rename the column "Weather" to "Weather_Condition."

In [24]:
# Renaming the col "Weather" to "Weather_Condition"
weather_df.rename(columns={"Weather": "Weather_Condition"}, inplace=True) 

# Confirming changes 
weather_df.head()

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   
1  1/1/2012 1:00    -1.8              -3.7         87                4   
2  1/1/2012 2:00    -1.8              -3.4         89                7   
3  1/1/2012 3:00    -1.5              -3.2         88                6   
4  1/1/2012 4:00    -1.5              -3.3         88                7   

   Visibility_km  Press_kPa     Weather_Condition  
0            8.0     101.24                   Fog  
1            8.0     101.24                   Fog  
2            4.0     101.26  Freezing Drizzle,Fog  
3            4.0     101.27  Freezing Drizzle,Fog  
4            4.8     101.23                   Fog

### Question 5: What is the mean visibility of the dataset?

In [28]:
# Getting mean visibility and rounding to 2 decimal places
mean_visibility = weather_df.Visibility_km.mean().round(2)

print("Mean Visibility: ", mean_visibility)

Mean Visibility:  27.66


### Question 6. Find the number of records where the wind speed is greater than 24 km/hr and visibility is equal to 25 km.

In [33]:
# FIlter based on the given condition
filtered_df = weather_df[(weather_df["Wind Speed_km/h"] > 24) & (weather_df["Visibility_km"] == 25)] 

# Get the number of records
num_records = len(filtered_df)
print(num_records)

308


### Question 7. What is the mean value of each column for each weather condition?

In [34]:
# Using describe to get mean for each column 
weather_df.describe()

Temp_C  Dew Point Temp_C    Rel Hum_%  Wind Speed_km/h  \
count  8784.000000       8784.000000  8784.000000      8784.000000   
mean      8.798144          2.555294    67.431694        14.945469   
std      11.687883         10.883072    16.918881         8.688696   
min     -23.300000        -28.500000    18.000000         0.000000   
25%       0.100000         -5.900000    56.000000         9.000000   
50%       9.300000          3.300000    68.000000        13.000000   
75%      18.800000         11.800000    81.000000        20.000000   
max      33.000000         24.400000   100.000000        83.000000   

       Visibility_km    Press_kPa  
count    8784.000000  8784.000000  
mean       27.664447   101.051623  
std        12.622688     0.844005  
min         0.200000    97.520000  
25%        24.100000   100.560000  
50%        25.000000   101.070000  
75%        25.000000   101.590000  
max        48.300000   103.650000

### 8. Find all instances where the weather is clear and the relative humidity is greater than 50, or visibility is above 40.

In [39]:
# Set conditions first
clear_weather_condition= weather_df["Weather_Condition"] == "Clear"
humidity_condition = weather_df["Rel Hum_%"] > 50 
visibility_condition = weather_df["Visibility_km"] > 40 

# Applying conditions
all_instances = weather_df[clear_weather_condition & humidity_condition | visibility_condition]
all_instances.head(10)

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
106  1/5/2012 10:00    -6.0             -10.0         73               17   
107  1/5/2012 11:00    -5.6             -10.2         70               22   
108  1/5/2012 12:00    -4.7              -9.6         69               20   
109  1/5/2012 13:00    -4.4              -9.7         66               26   
110  1/5/2012 14:00    -5.1             -10.7         65               22   
111  1/5/2012 15:00    -4.3             -12.0         55               26   
114  1/5/2012 18:00    -7.1             -14.4         56               11   
115  1/5/2012 19:00    -9.2             -15.4         61                7   
116  1/5/2012 20:00    -9.8             -15.7         62                9   
117  1/5/2012 21:00    -9.0             -14.8         63               13   

     Visibility_km  Press_kPa Weather_Condition  
106           48.3     100.45      Mainly Clear  
107           48.3     100.41      Mainly Clear  
108           48.3     100.38      Mainly Clear  
109           48.3     100.40      Mainly Clear  
110           48.3     100.46      Mainly Clear  
111           48.3     100.52      Mainly Clear  
114           25.0     100.71             Clear  
115           25.0     100.80             Clear  
116           25.0     100.83             Clear  
117           25.0     100.83             Clear